examples/react/src/hooks.js

```diff
@@ -0,0 +1,95 @@
+import { useEffect, useReducer } from "react";
+import { TodoModel } from './datastoreConfig';
+
+const Actions = {
+    REQ_START: 0,
+    REQ_SUCCESS: 1,
+    REQ_FAILED: 2
+}
+
+function reducer(state, action) {
+    switch (action.type) {
+        case Actions.REQ_START:
+            return { ...state, loading: true, data: null, error: null };
+        case Actions.REQ_SUCCESS:
+            return { ...state, loading: false, data: action.payload, error: null };
+        case Actions.REQ_FAILED:
+            return { ...state, loading: false, data: null, error: action.payload };
+        default:
+            throw new Error("Invalid action");
+    }
+}
+
+const initialState = {
+    data: null,
+    loading: false,
+    error: null
+};
+
+export const useFindTodos = () => {
+    const [state, dispatch] = useReducer(reducer, initialState);
+
+    useEffect(() => {
+        (async function () {
+            dispatch({ type: Actions.REQ_START });
+            try {
+                const result = await TodoModel.query();
+                dispatch({ type: Actions.REQ_SUCCESS, payload: result });
+            } catch (error) {
+                dispatch({ type: Actions.REQ_FAILED, payload: error });
+            }
+        })();
+    }, []);
+
+    return {
+        ...state, data: state.data ? state.data : []
+    }
+}
+
+export const useAddTodo = () => {
+    const [state, dispatch] = useReducer(reducer, initialState);
```

Comment:
Might be better to use context and do the `useReducer` only once, no?

This line could then become `const {state, dispatch} = useContext(CONTEXT_NAME);` and that way we are not re-creating the reducer multiple times with the `initialState`.

I would also almost be tempted to change things up here so that you store the model queries in the initialState. i.e. If you had `todos` and `comments`, your initial state could be 

```
const initialState = { todos: [], comments: [] }`
```

And then your reducer dispatch switch statements to handle the different CRUD types, eg:

```
async function reducer(state, action) {
    switch (action.type) {
        case Actions.FIND_TODOS:
			const todos = await TodoModel.query();
            return { ...state, loading: true, todos };
       	...
        default:
            throw new Error("Invalid action");
    }
}
```

This is just a suggestion. It may invalidate the need for some of the hooks though.

Reply:
Not too sure if you can do async reducer functions though.

Reply:
@kingsleyzissou are you talking about storing todos in a global context and then dispatching actions to it? If you are, how should users make unique queries with the useFindTodos hook? Note that, I can easily allow users to pass parameters to be used for querying into the hook state by changing the code a little.

Reply:
My concern is if we have an app that has 4 or 5 or n amounts of models, we will end up having a `find`, `create`, `update` and `delete` hook for each case. This would become unmanageable pretty quickly. Whereas a more generic hook, might work better. This is just a suggestion though as this is part of a sample app and not hooks that we are providing the developer.

Reply:
We can make the hooks more generic by passing the Model as an input parameter
`useAdd(TodoModel);`
This we can create only one hook for each operation that are reusable for all models

Reply:
My suggestion is to create issue with this details. `useAdd` or `useDataStore(Action.ADD)` or `useDataStore().add()` as options to talk about

Reply:
@Eunovo Can you create issue for this?